### ISSUE TITLE
Memory leak when handling exception on ubuntu 14.04

### ISSUE THREAD

**Issue by hpaiemen** [dotnet/coreclr#1420]
Here's the repro :

```c#
public void Main(string[] args)
{
      int i = 0;
      while(true)
      {
            try
            {
                throw new Exception();
            }
            catch(Exception e)
            {
                if (e != null)
                    i++;
            }
      }
}
```
Running this simple code with dnx on Linux produce a memory leak. I've monitored a leak around 20 MB/Min.

**Comment by ayende** [comment]
What happens if you run a GC at some point?

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by hpaiemen** [comment]
Does the same here's the code :
```c#
public void Main(string[] args)
        {
            int i = 0;
            while(true)
            {
                try
                {
                    throw new Exception();
                }
                catch(Exception e)
                {
                    if (e != null)
                        i++;
                }
                if (i % 10000 == 0)
                {
                    GC.Collect();
                    Console.WriteLine("Collected");
                }
            }
        }
```

**Comment by sergiy-k** [comment]
Thank you, @hpaiemen! I took a quick look and apparently there are a few memory leaks. First, the PAL_GetStackBase() and PAL_GetStackLimit() functions are leaking memory because they are not calling pthread_attr_destroy() for the thread attribute object. These can be easily fixed.

A more interesting issue here is that we are also leaking memory that c++ runtime allocates when we call “throw PAL_SEHException(...)” in the SEHRaiseException() function in PAL. The problem is that we never return from the catch statement for PAL_SEHException in exceptmacros.h and exceptionhandling.cpp in the vm code: the catch block calls DispatchManagedException() which never returns because if it finds a managed handler for the exception, the code will simply call RtlRestoreContext to resume execution of managed code.  As a result, c++ runtime does not get a chance to do the cleanup. You can easily see that if you take a look at disassembly for the catch block.

cc: @janvorli

**Comment by janvorli** [comment]
Thanks @sergiy-k for looking into the issue. I will fix all of these.

**Comment by janvorli** [comment]
Fixed by #1496
